tests/test_packages/test_connections/test_p2p_libp2p/test_fault_tolerance.py

```diff
@@ -71,174 +65,119 @@ def change_state_and_wait(
 
 
 @libp2p_log_on_failure_all
-class TestLibp2pConnectionRelayNodeRestartIncomingEnvelopes(BaseTestLibp2pRelay):
-    """Test that connection will reliably receive envelopes after its relay node restarted"""
-
-    @classmethod
-    def setup_class(cls):
-        """Set the test up"""
-        super().setup_class()
-        cls.genesis = _make_libp2p_connection()
-
-        cls.multiplexer_genesis = Multiplexer([cls.genesis], protocols=[DefaultMessage])
-        cls.multiplexer_genesis.connect()
-        cls.log_files.append(cls.genesis.node.log_file)
-        cls.multiplexers.append(cls.multiplexer_genesis)
-
-        genesis_peer = cls.genesis.node.multiaddrs[0]
-
-        cls.relay = _make_libp2p_connection(entry_peers=[genesis_peer])
-        cls.multiplexer_relay = Multiplexer([cls.relay], protocols=[DefaultMessage])
-        cls.multiplexer_relay.connect()
-        cls.log_files.append(cls.relay.node.log_file)
-        cls.multiplexers.append(cls.multiplexer_relay)
-
-        relay_peer = cls.relay.node.multiaddrs[0]
-
-        cls.connection = _make_libp2p_connection(relay=False, entry_peers=[relay_peer])
-        cls.multiplexer = Multiplexer([cls.connection], protocols=[DefaultMessage])
-        cls.multiplexer.connect()
-        cls.log_files.append(cls.connection.node.log_file)
-        cls.multiplexers.append(cls.multiplexer)
-
-        cls.connection2 = _make_libp2p_connection(relay=False, entry_peers=[relay_peer])
-        cls.multiplexer2 = Multiplexer([cls.connection2], protocols=[DefaultMessage])
-        cls.multiplexer2.connect()
-        cls.log_files.append(cls.connection2.node.log_file)
-        cls.multiplexers.append(cls.multiplexer2)
+class TestLibp2pConnectionRelayNodeRestart(BaseTestLibp2pRelay):
+    """Test that connection will reliably forward envelopes after its relay node restarted"""
+
+    def setup(self):
+        """Set up the individual test method"""
+
+        self.genesis = self.make_connection()
+        genesis_peer = self.genesis.node.multiaddrs[0]
+        self.relay = self.make_connection(entry_peers=[genesis_peer])
+        relay_peer = self.relay.node.multiaddrs[0]
+        self.connection1 = self.make_connection(relay=False, entry_peers=[relay_peer])
+        self.connection2 = self.make_connection(relay=False, entry_peers=[relay_peer])
+
+        # create references for disconnecting and reconnecting
+        self.multiplexer_genesis = self.multiplexers[0]
+        self.multiplexer_relay = self.multiplexers[1]
+        self.multiplexer1 = self.multiplexers[2]
+        self.multiplexer2 = self.multiplexers[3]
+
+    def teardown(self):
+        """Teardown"""
+        self.disconnect()
+        self.multiplexers.clear()
```

Comment:
All child classes (actual tests) should implement setup instead of setup_class in step of incremental refactoring. That is much better as individual tests should not share state needlessly (especially connections that are disconnected / connected, the order of execution of tests is not guaranteed either. I [alluded to this here](https://github.com/valory-xyz/open-aea/pull/231#discussion_r926819289)